[Feature] Allow structured grading instructions editable after first usage

imstreetspirit opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/server.html).
- [x] Server: I added multiple integration tests (Spring) related to the features (with a high test coverage)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Client: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/client.html).
- [x] Client: I added multiple integration tests (Jest) related to the features (with a high test coverage)
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
The structured grading instructions enable instructors and teaching assistants to provide standard feedback during the assessment. Artemis does not support to change structured grading instructions when the feedback of instructions is used in the assessment. In some cases, misspelling in the instructions required to be fixed during the assessment. Also, Instructors/TAs can not add new grading instruction during the assessment. As a result, the instructors and the teaching assistants have to write feedback manually and the instructions are not very useful for them. This PR provides a solution for the instructors, and they can edit the instructions in a more structured way.

### Description
<!-- Describe your changes in detail -->

Part I
To understand if the structured grading instruction is used or not used, `isGradingInstructionFeedbackUsed` is set for the exercise when the server sends exercise details to client side. If it is used, the grading instructions are displayed in form (similar with the re-evaluation of quiz exercises) 
`Feedback `and `GradingInstruction `have many-to-one relationship, so deleting only grading instruction would lead data integrity violation. Instructors could want that all corresponding feedback is deleted, but they could also want that only the association is removed and the feedback still stays valid. Our solution is when the grading instruction deleted, the association between `GradingInstruction `and `Feedback `is removed. The follow-up task could be some option to remove the all feedback which is associated with the grading instruction that will be deleted.

Part II (This will be solved in a follow up PR. Until then, user will be warned when he/she change the instruction credit or delete the instruction)
If the existing instruction which is associated with `Feedback` is updated by Instructor, the corresponding feedback should also updated. Re-evaluation of feedback is not implemented yet, it requires additional tasks. When it is done, the feedback text, credit would be updated and the result would be re-assessed. While the user is updating the exercise, Artemis would warn the instructor/editor that these changes have consequences for the existing assessments. 

### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->

1. Log in to Artemis
2. Create/Find a new exercise (could be any type of exercise except quiz)
3. Switch structured grading instruction, then save the exercise.
4. Add an example solution for the exercise / participate this exercise as a student
5. If you add example submission, then create new assessment for this submission and use one of the structured grading instruction as a feedback. (You can drag&drop the instruction into the feedback area. see >> screen record)
5. If you participated this exercise as student, then login as tutor/instructor and assess this submission. During the assessment, use one of the structured grading instructions as a feedback. (You can drag&drop the instruction into the feedback area. see >> screen record)
6. After that, as an instructor, go to edit view for this exercise.
7. switch the structured grading instruction button, after first usage of the feedback, the instructions should be displayed in form.
8. You can add/edit/delete grading criteria title or grading instructions. please test the all new buttons.
9. Then try to save the exercise, there will be no "Internal Server Error", and you can see updated grading instruction in the exercise detail page.
    a) if you delete or change the credit for the existing instructions, the warning pop-up should be displayed after you click "Save" button. 
       (see screenshot) The exercise should be updated when you click "Confirm Changes".



**You should also check that structured grading instructions can be edited when there is no submissions which means that grading instructions are not used as feedback.

***If you use grading instructions to asesss modeling exercise UML elements inside of the Apollon Editor (see last screenshot), currently it can not match the instruction with the feedback. So, wehen you go back to edit page, you can not see the grading instructions in a new form. It requires additional information from Apollon editor. This will be solved in a follow up.

### Test Coverage
<!-- Please add the test coverage for all changes files here. You can see this when executing the tests locally (see build.gradle and package.json) or when looking into the corresponding Bamboo build plan -->
<!-- * ExerciseService.java: 85% -->
<!-- * programming-exercise.component.ts 95% -->

grading-instructions-details.component.ts 59% (only new functions are  covered, will be increased in the follow-up task)
FileUploadExerciseResource.java 100%
TextExerciseResource.java 92%
ModelingExerciseResource.java 93%

### Screenshots
<!-- Add screenshots to demonstrate the changes in the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/2017/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->

Following view represents the current version of structured grading instruction.

<img src="https://user-images.githubusercontent.com/23650630/116827748-f5146580-ab9a-11eb-8314-38db26fc6182.png" width="400">

Following view represents the  structured grading instruction if one of them used as a feedback.

<img src="https://user-images.githubusercontent.com/23650630/116827802-4ae90d80-ab9b-11eb-8c94-6c6897775e26.png" width="400">

How to use structured grading instruction as feedback during the assessment period

https://user-images.githubusercontent.com/23650630/116828319-3c502580-ab9e-11eb-84ea-b48414edcbb1.mp4

The warning pop-up only if you change the instruction credit or delete one of the instruction/criteria which is associated with the feedback

<img src="https://user-images.githubusercontent.com/23650630/117378695-34fa8600-aed6-11eb-9af4-c8761b6c0e35.png" width="400">

Modeling Assessment for UML elements - user can drag&drop the single instruction into the UML element

<img src="https://user-images.githubusercontent.com/23650630/118262104-b47aed00-b4b4-11eb-9911-01f0de5b5108.png" width="400">

KiraNickel reviewed (approved):
Tested on ts1, works as expected 👍🏼 
The assessments that were already finished did not change and were not deleted after the grading Instructions were changed, which as I understood it is correct for the Part I which is covered in this issue.

There was one problem but I think it has been encountered before and is unrelated to this PR. When the instructions are changed, the cursor sometimes jumps to the end of the input box. Also, when the instructions are saved, the character which was entered last disappears.

imstreetspirit commented:
> Tested on ts1, works as expected 👍🏼
> The assessments that were already finished did not change and were not deleted after the grading Instructions were changed, which as I understood it is correct for the Part I which is covered in this issue.
> 
> There was one problem but I think it has been encountered before and is unrelated to this PR. When the instructions are changed, the cursor sometimes jumps to the end of the input box. Also, when the instructions are saved, the character which was entered last disappears.

Thank your for your review 😊 The current PR only covers changing the instructions. As a follow-up, the feedback scores will be updated in the assessments, and the submissions should be re-evaluated.

hmm, I'll look at the cursor problem, this could be related with the markdown editor, need to check that. 👍

real-dmy reviewed (commented):
Tested on TS1. Created one text and one model exercise. 
Text exercise worked just fine(I added an example submission and changed the grading criteria afterwards.) One problem was that after I have reviewed the example submission and wanted to change the 'credit' in the 'structured Grading Instructions', immediately after I write a number 'NaN' was written.
![image](https://user-images.githubusercontent.com/80687511/117567949-b7459e80-b0be-11eb-8522-5d880a4c5f95.png)

Model exercise: I have created an example submission and graded it with the structured grading criteria. But when I went on to the edit mode and switched to Structured Grading Instructions, the wanted view did not appear. The form was the same as before(see picture).
![image](https://user-images.githubusercontent.com/80687511/117567836-47cfaf00-b0be-11eb-985f-88371b303396.png)

imstreetspirit commented:
> Tested on TS1. Created one text and one model exercise.
> Text exercise worked just fine(I added an example submission and changed the grading criteria afterwards.) One problem was that after I have reviewed the example submission and wanted to change the 'credit' in the 'structured Grading Instructions', immediately after I write a number 'NaN' was written.
> ![image](https://user-images.githubusercontent.com/80687511/117567949-b7459e80-b0be-11eb-8522-5d880a4c5f95.png)
> 
> Model exercise: I have created an example submission and graded it with the structured grading criteria. But when I went on to the edit mode and switched to Structured Grading Instructions, the wanted view did not appear. The form was the same as before(see picture).
> ![image](https://user-images.githubusercontent.com/80687511/117567836-47cfaf00-b0be-11eb-985f-88371b303396.png)

Thank you for the review 😊 I'll look at the issue for the editor. 
 I just check the database and when feedbacks are submitted for modeling exercise assessment, the association with grading instruction set null. So when you go to edit page, it can not find the instructions which is associated with the feedback used. Thank you for noticing. I need to ask this if it is intended behavior or there is something overlooked.

canyelten reviewed (dismissed):
Tested on TS1. I couldn't find any other problem than the mentioned "NaN" issue, where the cursor sometimes jumps to the end of the input box.

willergeorg reviewed (commented):
Tested on TS2. Besides the problems mentioned earlier, everything worked.

FrankeLukas commented on src/main/java/de/tum/in/www1/artemis/repository/FeedbackRepository.java:
```diff
@@ -153,6 +158,18 @@ else if (!testMessages.isEmpty()) {
+     * @return Feedback list which are associated with the grading instructions
+     */
+    default List<Feedback> findFeedbackByStructuredGradingInstructionId(List<GradingCriterion> gradingCriteria) {
+        List<Long> list = gradingCriteria.stream().flatMap( gradingCriterion ->  gradingCriterion.getStructuredGradingInstructions().stream()).map(GradingInstruction::getId).collect(toList());
+        return findFeedbackCountByStructuredGradingInstructionIds(list);
```
```suggestion
        List<Long> gradingInstructionsIds = gradingCriteria.stream().flatMap(gradingCriterion -> gradingCriterion.getStructuredGradingInstructions().stream()).map(GradingInstruction::getId).collect(toList());
        return findFeedbackCountByStructuredGradingInstructionIds(gradingInstructionsIds);
```

FrankeLukas commented on src/main/java/de/tum/in/www1/artemis/repository/FeedbackRepository.java:
```diff
@@ -40,6 +42,9 @@
 
     List<Feedback> findByReferenceInAndResult_Submission_Participation_Exercise(List<String> references, Exercise exercise);
 
+    @Query("select feedback from Feedback feedback where feedback.gradingInstruction.id in :gradingInstructionsId")
+    List<Feedback> findFeedbackCountByStructuredGradingInstructionIds(@Param("gradingInstructionsId") List<Long> gradingInstructionsId);
```
```suggestion
    @Query("select feedback from Feedback feedback where feedback.gradingInstruction.id in :gradingInstructionsIds")
    List<Feedback> findFeedbackCountByStructuredGradingInstructionIds(@Param("gradingInstructionsIds") List<Long> gradingInstructionsIds);
```

imstreetspirit replied:
resolved.

FrankeLukas commented on src/main/java/de/tum/in/www1/artemis/web/rest/FileUploadExerciseResource.java:
```diff
@@ -254,6 +257,12 @@ else if (!authCheckService.isAtLeastTeachingAssistantForExercise(fileUploadExerc
 
+        List<Feedback> feedbackList = feedbackRepository.findFeedbackByStructuredGradingInstructionId(gradingCriteria);
+
+        if(!feedbackList.isEmpty()) {
```
```suggestion
        if (!feedbackList.isEmpty()) {
```

FrankeLukas commented on src/main/java/de/tum/in/www1/artemis/web/rest/ModelingExerciseResource.java:
```diff
@@ -255,6 +258,11 @@ public ModelingExerciseResource(ModelingExerciseRepository modelingExerciseRepos
         List<GradingCriterion> gradingCriteria = gradingCriterionRepository.findByExerciseIdWithEagerGradingCriteria(exerciseId);
+        List<Feedback> feedbackList = feedbackRepository.findFeedbackByStructuredGradingInstructionId(gradingCriteria);
+
+        if(!feedbackList.isEmpty()) {
```
```suggestion
        if (!feedbackList.isEmpty()) {
```

imstreetspirit replied:
resolved.

FrankeLukas commented on src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java:
```diff
@@ -675,6 +679,12 @@ private void validateStaticCodeAnalysisSettings(ProgrammingExercise programmingE
+
+        List<Feedback> feedbackList = feedbackRepository.findFeedbackByStructuredGradingInstructionId(gradingCriteria);
+
+        if(!feedbackList.isEmpty()) {
```
```suggestion
        if (!feedbackList.isEmpty()) {
```

imstreetspirit replied:
resolved.

FrankeLukas commented on src/main/java/de/tum/in/www1/artemis/web/rest/TextExerciseResource.java:
```diff
@@ -293,6 +293,12 @@ else if (!authCheckService.isAtLeastTeachingAssistantForExercise(optionalTextExe
+
+        List<Feedback> feedbackList = feedbackRepository.findFeedbackByStructuredGradingInstructionId(gradingCriteria);
+
+        if(!feedbackList.isEmpty()) {
```
```suggestion
        if (!feedbackList.isEmpty()) {
```

FrankeLukas commented on src/main/webapp/i18n/de/exercise.json:
```diff
@@ -172,7 +174,18 @@
+                    "warning": "Warnung!",
+                    "confirm": "Änderungen bestätigen",
+                    "problems": "Folgende Probleme können auftreten:",
+                    "instruction_deleted": "Benotungshinweise gelöscht:",
+                    "scoring_change": "Bewertungsart verändert:",
```
```suggestion
                    "instructionDeleted": "Benotungshinweise gelöscht:",
                    "scoringChange": "Bewertungsart verändert:",
```

imstreetspirit replied:
resolved.

FrankeLukas commented on src/main/webapp/i18n/en/exercise.json:
```diff
@@ -172,7 +174,18 @@
+                    "warning": "Warning!",
+                    "confirm": "Confirm changes",
+                    "problems": "The following problems may occur:",
+                    "instruction_deleted": "Grading Instruction deleted:",
+                    "scoring_change": "Scoring changed:",
```
```suggestion
                    "instructionDeleted": "Benotungshinweise gelöscht:",
                    "scoringChange": "Bewertungsart verändert:",
```

imstreetspirit replied:
resolved.

FrankeLukas reviewed (changes requested):
A few style remarks

NicolasUllmann reviewed (approved):
Tested on Server 1. Concerning the functionality, everything worked as expected, except for the remark mentioned above.

imstreetspirit reviewed (commented):
The 'NaN' issue and the cursor issue in Markdown should be resolved.

PatZintel reviewed (approved):
Found no problem on ts2, just one small comfort change:
When clicking on the plus for adding a new Grading instruction, the new Instruction gets spawned with empty textfield, i think it would make sense to have the default template for the instructions just like when you start creating new Instructions.

lschlesinger commented on src/main/webapp/app/exercises/shared/structured-grading-criterion/grading-instructions-details/grading-instructions-details.component.html:
```diff
@@ -1,40 +1,108 @@
-                        </tr>
-                    </tbody>
-                </table>
+<div *ngIf="!exercise.gradingInstructionFeedbackUsed">
```
Why does this look different to the view when editing the exercise before having grading instruction feedback used?
![image](https://user-images.githubusercontent.com/41366522/117764796-55f00d80-b22d-11eb-9341-bdf9d5ab930b.png) vs. 
![image](https://user-images.githubusercontent.com/41366522/117764850-67d1b080-b22d-11eb-8824-eff483091535.png)

I get why you need to let the user confirm changes after having already used the feedback, but beside, from a user perspective I would expect the edit interface to be the same.

iozbeyli replied:
Also refresh button next to inputs are not aligned

imstreetspirit replied:
> Also refresh button next to inputs are not aligned 
resolved.

To Lorena>>
hmm, we decided to change view only if the instructions are used. Maybe the existing view may change in the future, for now we have decided to keep it. I took a note, I'll bring it up in the meeting... I expect some feedbacks about that after being used.

lschlesinger commented on src/main/webapp/app/exercises/shared/structured-grading-criterion/grading-instructions-details/grading-instructions-details.component.html:
```diff
@@ -1,40 +1,108 @@
+                                    <fa-icon [icon]="'trash'" size="2x"></fa-icon>
+                                </span>
+                                <span class="re-evaluate-button" id="add-button" (click)="addNewInstruction(criteria)">
+                                    <fa-icon [icon]="'plus'" size="2x"></fa-icon>
```
Do we ever use icons of size 2x? I would stay with the 1x size, but that is personal taste I guess :)

imstreetspirit replied:
👍 Personally, I like the minimal design. I mostly copied the design from the re-evaluation of quizzes (added screenshot). I just thought that if this is approved for quiz re-evaluation, I can use it for my work. If you think that small size fits better, i can change it quickly.

<img src="https://user-images.githubusercontent.com/23650630/117786297-a2dfde00-b245-11eb-80fb-3abd2ecd053b.png" width="400"></img>

lschlesinger reviewed (changes requested):
I am not sure if I did something wrong, but following you instructions, I ran in to the following situation:
I edited a instruction that was already used (changed points from 0 to 9), confirmed the changes and in the details view, these changes were not reflected - but they should, I guess?

https://user-images.githubusercontent.com/41366522/117765166-dadb2700-b22d-11eb-9393-c6dc443d9f19.mov

iozbeyli commented on src/main/webapp/app/exercises/shared/exercise-update-warning/exercise-update-warning.component.ts:
```diff
@@ -0,0 +1,33 @@
+
+    constructor(public activeModal: NgbActiveModal) {}
+
+    ngOnInit(): void {}
```
No need for this

iozbeyli commented on src/main/webapp/app/exercises/shared/exercise-update-warning/exercise-update-warning.component.html:
```diff
@@ -0,0 +1,36 @@
+        <ul>
+            <li *ngIf="instructionDeleted" style="padding-top: 5px">
+                <span jhiTranslate="artemisApp.exercise.update.warning.instructionDeleted"></span>
+                <ul style="padding-left: 15px">
+                    <li style="margin-bottom: -10px">
```
New css classes can be added so you can change both instruction deleted and scoring changed at the same time

iozbeyli commented on src/main/webapp/app/exercises/shared/structured-grading-criterion/grading-instructions-details/grading-instructions-details.component.scss:
```diff
@@ -0,0 +1,60 @@
+        border-radius: 0.1em;
+        border-width: 1px;
+        border-style: solid;
+        border-color: lightgrey;
```
lightgray would be better here. https://www.w3schools.com/colors/colors_shades.asp

imstreetspirit replied:
is there any difference between `lightgrey` and `lightgray`?

iozbeyli replied:
lightgray is more generally supported

iozbeyli commented on src/main/webapp/app/exercises/shared/structured-grading-criterion/grading-instructions-details/grading-instructions-details.component.ts:
```diff
@@ -285,4 +320,113 @@ export class GradingInstructionsDetailsComponent implements OnInit {
+     * @param {GradingCriterion} criteria
+     */
+    onInstructionChange(domainCommands: [string, DomainCommand | null][], instruction: GradingInstruction): void {
+        this.instructions = [];
+        this.instructions.push(instruction);
```
Why not just put inside it?

imstreetspirit replied:
want to be sure following `setInstructionParameters` runs for only this instruction. so, i just initialized with an empty array every time `onInstructionChange` is called

iozbeyli commented on src/main/webapp/app/exercises/shared/structured-grading-criterion/grading-instructions-details/grading-instructions-details.component.ts:
```diff
@@ -285,4 +320,113 @@ export class GradingInstructionsDetailsComponent implements OnInit {
+        const backupCriteriaIndex = this.backupExercise.gradingCriteria!.findIndex((gradingCriteria) => {
+            return gradingCriteria.id === criteria.id;
+        });
+        const backupInstructionIndex = this.backupExercise.gradingCriteria![criteriaIndex].structuredGradingInstructions.findIndex((sgi) => {
```
Shouldn't this be [backupCriteriaIndex]? I don't know the code but I am asking to be sure

iozbeyli commented on src/main/webapp/app/exercises/shared/structured-grading-criterion/grading-instructions-details/grading-instructions-details.component.ts:
```diff
@@ -285,4 +320,113 @@ export class GradingInstructionsDetailsComponent implements OnInit {
+     * @param criteria {GradingCriterion} the criteria, which includes title that will be reset
+     */
+    resetCriteriaTitle(criteria: GradingCriterion) {
+        const criteriaIndex = this.exercise.gradingCriteria!.findIndex((gradingCriteria) => {
+            return gradingCriteria.id === criteria.id;
+        });
```
You use index finding a lot for these grading criteria so it can be turned into a function

iozbeyli commented on src/main/webapp/app/exercises/shared/structured-grading-criterion/structured-grading-criterion.service.ts:
```diff
@@ -32,23 +32,23 @@ export class StructuredGradingCriterionService {
                 if (encounters >= maxCount) {
-                    gradingInstructions[feedback.gradingInstruction!.id] = encounters + 1;
+                    gradingInstructions[feedback.gradingInstruction!.id!] = encounters + 1;
                 } else {
-                    gradingInstructions[feedback.gradingInstruction!.id] = encounters + 1;
+                    gradingInstructions[feedback.gradingInstruction!.id!] = encounters + 1;
```
If else has the same code so why the repetition?

iozbeyli replied:
Same thing happens inside both if and else statements. I don't know if you are going to do anything about it since you resolved it. I just wanted to point it out again in case you forgot it.

ArtemTsurikov reviewed (changes requested):
Tested on ts1. All functionalities worked. The "Delete This Criteria" Button look a little bit weird on my 12inch screen.
<img width="1101" alt="Screenshot 2021-05-11 at 12 05 57" src="https://user-images.githubusercontent.com/64026687/117799209-24d60400-b252-11eb-8d40-e59b0545e063.png">

imstreetspirit commented:
> I am not sure if I did something wrong, but following you instructions, I ran in to the following situation:
> I edited a instruction that was already used (changed points from 0 to 9), confirmed the changes and in the details view, these changes were not reflected - but they should, I guess?
> 
>  update-grading-instructions.mov

willergeorg reviewed (approved):
Tested again on TS1. Handling was smooth, no NaNs or unexpected jumps of the cursor when trying to edit an instruction. Re-approve.

NicolasUllmann reviewed (dismissed):
Tested on Server 1. Everything worked as expected.

lschlesinger reviewed (approved):
I tested it on TS1. Besides the things you want to tackle in the follow-up PR (especially the recalculation of criteria already used as feedback if edited), I did not encounter any problems.

MDK5 commented on src/main/java/de/tum/in/www1/artemis/repository/FeedbackRepository.java:
```diff
@@ -40,6 +42,9 @@
     List<Feedback> findByReferenceInAndResult_Submission_Participation_Exercise(List<String> references, Exercise exercise);
 
+    @Query("select feedback from Feedback feedback where feedback.gradingInstruction.id in :gradingInstructionsIds")
+    List<Feedback> findFeedbackCountByStructuredGradingInstructionIds(@Param("gradingInstructionsIds") List<Long> gradingInstructionsIds);
```
This name suggests it would return a number ("feedbackCount"), "findFeedbacksBy..." would probably be more intuitive. Or I missunderstood what the method does.

imstreetspirit replied:
thank you for noticing, I was using count before, I just forgot to change method name after returning the Feedback list.

MDK5 commented on src/main/webapp/app/exercises/shared/structured-grading-criterion/grading-instructions-details/grading-instructions-details.component.html:
```diff
@@ -1,40 +1,107 @@
+                    <table class="table table-bordered">
+                        <thead>
+                            <tr>
+                                <td class="font-weight-bold">Credits</td>
```
I would expect this to be translated as well

imstreetspirit replied:
oh well, I did not focus on the existing structure, you're right, this one should be translated 😉

imstreetspirit replied:
I guess `credits ` and `feedback ` are same for English and German, that's why they are not included in the translation files at first. But I am a beginner when it comes to German 😄

MDK5 replied:
I see. I would've translated it with "Punkte", but now I realise credits are referred to as points on other places in Artemis (most notably the points selector in the exercise creation) and I wonder why they are referred to as credits here (but are still the same thing from my understanding). I agree this is out of scope for this PR though.

MDK5 commented on src/main/webapp/i18n/en/exercise.json:
```diff
@@ -45,6 +45,8 @@
             "gradingInstructions": "Grading Instructions",
             "structuredGradingInstructions": "Structured Grading Instructions",
+            "addGradingCriteria": "Add Grading Criteria",
+            "deleteGradingCriteria": "Delete This Criteria",
```
I think it should be either singular "Delete This Criterion" or plural "Delete These Criteria"

MDK5 reviewed (approved):
Reviewed the code, changes look good to me from my limited typescript knowledge, I only found a few details that struck me odd.

KiraNickel reviewed (dismissed):
Tested on ts3 with a text exercise. Everything works as described and there are no more problems with the editor swallowing characters or the cursor skipping either 👍🏼

MDK5 reviewed (approved):
Re-Approve, code looks good to me

KiraNickel reviewed (approved):
Re-approve, tested on ts1, everything still works as expected 👍🏼

canyelten reviewed (approved):
Tested on TS1. Works as expected, re-approve.

ndricimrr commented:
I tested _Locally_ . Other than these two comments below the rest works as expected. 
First, when I first click on the **Structured Grading Instructions** switch in the _Text Exercise Edit Page_ I get an error printed in the Console. Video shown below 

https://user-images.githubusercontent.com/18335489/118394924-f4260e00-b647-11eb-999f-6a646b61aa5e.mp4

- On another note, most likely I did not understand correctly, but I am a bit confused by the two delete buttons. I am not sure what is the purpose of the Delete icon and and the red **Delete This Criterion** button there. What does one do, that the other doesn't. In the video below it looks like I can delete grading instructions but the title will stay. Is that expected ?

https://user-images.githubusercontent.com/18335489/118394992-449d6b80-b648-11eb-8ee4-faa56de66ea5.mp4

imstreetspirit commented:
> I tested _Locally_ . Other than these two comments below the rest works as expected.
> First, when I first click on the **Structured Grading Instructions** switch in the _Text Exercise Edit Page_ I get an error printed in the Console. Video shown below
> 
>  merve_pr_record.mp4 
> * On another note, most likely I did not understand correctly, but I am a bit confused by the two delete buttons. I am not sure what is the purpose of the Delete icon and and the red **Delete This Criterion** button there. What does one do, that the other doesn't. In the video below it looks like I can delete grading instructions but the title will stay. Is that expected ?
> 
>  merve_pr_record_2.mp4

okay, first one could be related with AfterInitView hooks. I do not have so much experience with Angular lifecycle hooks, I need to look at them in more detail. For the second one, we just wanted to design this part similar to re-evaluation of quiz exercise. So user can remove the instructions independently from the criteria. If the instructor wants to remove criteria completely, 'Delete this Criterion' provides this option. For now, criteria can exist without instruction, there is no mechanism to avoid that. However, I can ask this.

ndricimrr reviewed (approved):
Tested again and the [issue](https://github.com/ls1intum/Artemis/pull/3294#issuecomment-841802882) is fixed. Works as expected 👍

lschlesinger commented on src/main/webapp/app/exercises/shared/structured-grading-criterion/grading-instructions-details/grading-instructions-details.component.html:
```diff
@@ -1,40 +1,113 @@
+                                placeholder="Criterion Title"
+                            />
+                            <span class="col-auto input-group-btn d-flex align-items-center" id="reset-criteria-button" (click)="resetCriterionTitle(criteria)">
+                                <fa-icon [icon]="'undo'"></fa-icon>
```
IMO, You should add a `cursor: pointer;` css declaration for this icon.

imstreetspirit replied:
oh, cool.. that's good to know 👍

lschlesinger reviewed (approved):
Tested on TS1. It worked as described. I like the better embedding of the instructions in the exercise detail view!
I also did not encounter the problem mentioned by @ndricimrr.
I made a suggested as inline comment ;)

katjanakosic reviewed (approved):
Tested on TS2. Worked as expected, no errors.

iozbeyli commented on src/test/javascript/spec/service/exercise-update-warning.service.spec.ts:
```diff
@@ -0,0 +1,51 @@
+describe('Exercise Update Warning Service', () => {
+    let updateWarningService: ExerciseUpdateWarningService;
+
+    const gradingInstruction = { id: 1, credits: 1, gradingScale: 'scale', instructionDescription: 'description', feedback: 'feedback', usageCount: 0 } as GradingInstruction;
+    const gradingInstructionCreditsChanged = {
+        id: 1,
+        credits: 3,
+        gradingScale: 'scale',
+        instructionDescription: 'description',
+        feedback: 'feedback',
+        usageCount: 0,
+    } as GradingInstruction;
+    const gradingCriterion = { id: 1, title: 'testCriteria', structuredGradingInstructions: [gradingInstruction] } as GradingCriterion;
+    const gradingCriterionCreditsChanged = { id: 1, title: 'testCriteria', structuredGradingInstructions: [gradingInstructionCreditsChanged] } as GradingCriterion;
+    const gradingCriterionWithoutInstruction = { id: 1, title: 'testCriteria' } as GradingCriterion;
```
The spread operator might be helpful here to indicate that only credits is different between gradingInstruction and gradingInstructionCreditsChanged. Same thing for gradingCriterion. It also reduces code repetition.

iozbeyli commented on src/test/javascript/spec/component/shared/grading-instructions-details.component.spec.ts:
```diff
@@ -91,4 +99,70 @@ describe('Grading Instructions Management Component', () => {
     });
+
+    it('should return grading criteria index', () => {
+        component.exercise.gradingCriteria = [gradingCriterion];
```
This can be moved into before each it is repeated in all tests below.

iozbeyli commented on src/main/webapp/app/exercises/shared/structured-grading-criterion/grading-instructions-details/grading-instructions-details.component.ts:
```diff
@@ -285,4 +325,130 @@ export class GradingInstructionsDetailsComponent implements OnInit {
+     */
+    onInstructionChange(domainCommands: [string, DomainCommand | null][], instruction: GradingInstruction): void {
+        this.instructions = [];
+        this.instructions.push(instruction);
```
Why not just this.instructions = [instruction] ?

imstreetspirit replied:
yeah why not :) okay

iozbeyli commented on src/main/webapp/app/exercises/file-upload/manage/file-upload-exercise-update.component.ts:
```diff
@@ -77,10 +84,27 @@ export class FileUploadExerciseUpdateComponent implements OnInit {
         navigateBackFromExerciseUpdate(this.router, this.fileUploadExercise);
     }
 
+    save() {
+        if (this.fileUploadExercise.gradingInstructionFeedbackUsed) {
+            const ref = this.popupService.checkExerciseBeforeUpdate(this.fileUploadExercise, this.backupExercise);
+            if (!this.modalService.hasOpenModals()) {
+                this.saveExercise();
+            } else {
+                ref.then((reference) => {
+                    reference.componentInstance.confirmed.subscribe(() => {
+                        this.saveExercise();
+                    });
+                });
+            }
+        } else {
+            this.saveExercise();
+        }
+    }
```
Can this be moved into a service method since the same code is repeated in modeling-exercise-update?

imstreetspirit replied:
this part will changed for second part, I think i can move that part into service.

iozbeyli commented on src/main/java/de/tum/in/www1/artemis/web/rest/FileUploadExerciseResource.java:
```diff
@@ -254,6 +257,12 @@ else if (!authCheckService.isAtLeastTeachingAssistantForExercise(fileUploadExerc
         List<GradingCriterion> gradingCriteria = gradingCriterionRepository.findByExerciseIdWithEagerGradingCriteria(exerciseId);
         fileUploadExercise.setGradingCriteria(gradingCriteria);
 
+        List<Feedback> feedbackList = feedbackRepository.findFeedbackByStructuredGradingInstructionId(gradingCriteria);
+
+        if (!feedbackList.isEmpty()) {
+            fileUploadExercise.setGradingInstructionFeedbackUsed(true);
+        }
```
Do you have a service that you can move this since the same code appears for each exercise type?

imstreetspirit replied:
same thing for this part. I am defining new service methods for the second part, I'll use service for this work also.

ArtemTsurikov reviewed (approved):
Tested on TS1, looks good!
